crates/xilem_html/src/element/attributes.rs

```diff
@@ -0,0 +1,126 @@
+        match self {
+            AttributeValue::True => "".into(), // empty string is equivalent to a true set attribute
+            AttributeValue::I32(n) => n.to_string().into(),
+            AttributeValue::U32(n) => n.to_string().into(),
+            AttributeValue::F32(n) => n.to_string().into(),
+            AttributeValue::F64(n) => n.to_string().into(),
+            AttributeValue::String(s) => s.clone(),
+        }
+    }
+}
+
+pub trait IntoAttributeValue: Sized {
+    fn into_attribute_value(self) -> Option<AttributeValue>;
+}
+
+impl<T: IntoAttributeValue> IntoAttributeValue for Option<T> {
+    fn into_attribute_value(self) -> Option<AttributeValue> {
+        if let Some(value) = self {
+            T::into_attribute_value(value)
+        } else {
+            None
+        }
+    }
+}
+
+impl IntoAttributeValue for bool {
+    fn into_attribute_value(self) -> Option<AttributeValue> {
+        self.then_some(AttributeValue::True)
+    }
+}
+
+impl IntoAttributeValue for AttributeValue {
+    fn into_attribute_value(self) -> Option<AttributeValue> {
+        Some(self)
+    }
+}
+
+impl IntoAttributeValue for u32 {
+    fn into_attribute_value(self) -> Option<AttributeValue> {
+        Some(AttributeValue::U32(self))
+    }
+}
+
+impl IntoAttributeValue for i32 {
+    fn into_attribute_value(self) -> Option<AttributeValue> {
+        Some(AttributeValue::I32(self))
+    }
+}
+
+impl IntoAttributeValue for f32 {
+    fn into_attribute_value(self) -> Option<AttributeValue> {
+        Some(AttributeValue::F32(self))
+    }
+}
+
+impl IntoAttributeValue for f64 {
+    fn into_attribute_value(self) -> Option<AttributeValue> {
+        Some(AttributeValue::F64(self))
+    }
+}
+
+impl IntoAttributeValue for String {
+    fn into_attribute_value(self) -> Option<AttributeValue> {
+        Some(AttributeValue::String(self.into()))
+    }
+}
+
+impl IntoAttributeValue for CowStr {
+    fn into_attribute_value(self) -> Option<AttributeValue> {
+        Some(AttributeValue::String(self))
+    }
+}
+
+impl IntoAttributeValue for &'static str {
+    fn into_attribute_value(self) -> Option<AttributeValue> {
+        Some(AttributeValue::String(self.into()))
+    }
+}
+
+#[derive(Default)]
+pub struct Attributes(VecMap<CowStr, AttributeValue>);
+
+impl<'a> IntoIterator for &'a Attributes {
+    type Item = (&'a CowStr, &'a AttributeValue);
+
+    type IntoIter = <&'a VecMap<CowStr, AttributeValue> as std::iter::IntoIterator>::IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+impl Attributes {
+    // TODO return the previous attribute as an Option?
+    pub fn insert(&mut self, name: impl Into<CowStr>, value: impl IntoAttributeValue) {
+        let value = value.into_attribute_value();
+        // This is a simple optimization in case this is the first attribute inserted to the map (saves an allocation for the Vec)
+        if let Some(value) = value.into_attribute_value() {
```

Comment:
Double `.into_attribute_value()`. Also, is the comment still accurate?

Finally, how about providing `.remove(name)` as a more obvious alternative to `.insert(name, false)` (which I think makes a lot of sense to work, though).

Reply:
Whoops, I guess I was too fast pushing and overlooked that, thanks for catching this...

Yeah without the `Null`, a remove makes more sense probably, and handling that a layer above.
I think the whole wrapper type (`Attributes`) doesn't make much sense anymore (it was a wrapper around `VecMap` precisely because of guarding the potential `Null` value that would've been inserted).

I removed the wrapper type and renamed the file to `attribute_value.rs` as well, as it seems more appropriate in that case.